## Работа с данными в Pandas

In [1]:
import numpy as np
import pandas as pd

pd.set_option("display.precision", 2)

In [6]:
DATA_PATH = "telecom_churn.csv"

### Основные методы Pandas

Мы демонстрируем основные методы в действии, анализируя набор данных о показателях оттока клиентов операторов связи. Давайте прочитаем данные (используя метод read_csv) и посмотрим на первые 5 строк, используя метод head:

In [50]:
df = pd.read_csv(DATA_PATH)
df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

Посмотрим на ее размерность

In [8]:
print(df.shape)

(3333, 20)


Видно, что мы имеем дело с 20 столбцами и 3333 строками

Теперь посмотрим на названия столбцов в датасете

In [9]:
print(df.columns)

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')


Выведем общую информацию о столбцах.

К общей информации относится: индекс, название столбцов, кол-во не нулевых элементов в наборе и тип данных в столбце.

Важно отметить, что елси тип данных во всем столбце невозможно однозначно определить, то столбцу назначается тип Object.

In [34]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   bool   
 4   Voice mail plan         3333 non-null   bool   
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

Теперь давайте преобразуем тип в рамках столбца. К примеру из bool переведем в int64.

In [51]:
df["Churn"] = df["Churn"].astype("int64")

Проверим этот факт в информации о таблице.

In [13]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

#### Таким нехитрым способом можно вывести общие статистики вещественных фичей по столбцам во всей таблице

In [12]:
df.describe()

Account length  Area code  Number vmail messages  Total day minutes  \
count         3333.00    3333.00                3333.00            3333.00   
mean           101.06     437.18                   8.10             179.78   
std             39.82      42.37                  13.69              54.47   
min              1.00     408.00                   0.00               0.00   
25%             74.00     408.00                   0.00             143.70   
50%            101.00     415.00                   0.00             179.40   
75%            127.00     510.00                  20.00             216.40   
max            243.00     510.00                  51.00             350.80   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count          3333.00           3333.00            3333.00          3333.00   
mean            100.44             30.56             200.98           100.11   
std              20.07              9.26              50.71            19.92   
min               0.00              0.00               0.00             0.00   
25%              87.00             24.43             166.60            87.00   
50%             101.00             30.50             201.40           100.00   
75%             114.00             36.79             235.30           114.00   
max             165.00             59.64             363.70           170.00   

       Total eve charge  Total night minutes  Total night calls  \
count           3333.00              3333.00            3333.00   
mean              17.08               200.87             100.11   
std                4.31                50.57              19.57   
min                0.00                23.20              33.00   
25%               14.16               167.00              87.00   
50%               17.12               201.20             100.00   
75%               20.00               235.30             113.00   
max               30.91               395.00             175.00   

       Total night charge  Total intl minutes  Total intl calls  \
count             3333.00             3333.00           3333.00   
mean                 9.04               10.24              4.48   
std                  2.28                2.79              2.46   
min                  1.04                0.00              0.00   
25%                  7.52                8.50              3.00   
50%                  9.05               10.30              4.00   
75%                 10.59               12.10              6.00   
max                 17.77               20.00             20.00   

       Total intl charge  Customer service calls    Churn  
count            3333.00                 3333.00  3333.00  
mean                2.76                    1.56     0.14  
std                 0.75                    1.32     0.35  
min                 0.00                    0.00     0.00  
25%                 2.30                    1.00     0.00  
50%                 2.78                    1.00     0.00  
75%                 3.27                    2.00     0.00  
max                 5.40                    9.00     1.00

Также можно посмотреть разные статистики отдельно по столбцам

In [39]:
df["Churn"].mean(), df["Churn"].min(), df["Churn"].max(), df["Churn"].std(), df["Churn"].median()

(0.14491449144914492, 0, 1, 0.35206742362412946, 0.0)

Если задать уточнение типов, которые нас интересуют, то можно собрать более узкий срез статистик

In [14]:
df.describe(include=["object", "bool"])

State International plan Voice mail plan
count   3333               3333            3333
unique    51                  2               2
top       WV                 No              No
freq     106               3010            2411

Также можем посмотреть на количество уникальных элементов в столбце(гистограмма)

In [15]:
df["Churn"].value_counts()

Churn
0    2850
1     483
Name: count, dtype: int64

Дополнительным параметром `normalize=True` можно указать нормализовать переменные.

In [16]:
df["Churn"].value_counts(normalize=True)

Churn
0    0.86
1    0.14
Name: proportion, dtype: float64

### Сортировка

Отсортируем таблицу по столбцу `Total day charge` в порядке невозрастания.

In [17]:
df.sort_values(by="Total day charge", ascending=False).head()

State  Account length  Area code International plan Voice mail plan  \
365     CO             154        415                 No              No   
985     NY              64        415                Yes              No   
2594    OH             115        510                Yes              No   
156     OH              83        415                 No              No   
605     MO             112        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
365                       0              350.8               75   
985                       0              346.8               55   
2594                      0              345.3               81   
156                       0              337.4              120   
605                       0              335.5               77   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
365              59.64              216.5               94             18.40   
985              58.96              249.5               79             21.21   
2594             58.70              203.4              106             17.29   
156              57.36              227.4              116             19.33   
605              57.04              212.5              109             18.06   

      Total night minutes  Total night calls  Total night charge  \
365                 253.9                100               11.43   
985                 275.4                102               12.39   
2594                217.5                107                9.79   
156                 153.9                114                6.93   
605                 265.0                132               11.93   

      Total intl minutes  Total intl calls  Total intl charge  \
365                 10.1                 9               2.73   
985                 13.3                 9               3.59   
2594                11.8                 8               3.19   
156                 15.8                 7               4.27   
605                 12.7                 8               3.43   

      Customer service calls  Churn  
365                        1      1  
985                        1      1  
2594                       1      1  
156                        0      1  
605                        2      1

Также можно отсортировать таблицу по нескольким столбцам в определенном порядке для каждого. Для этого необходимо подать два списка в порядках приоритета. В данном случае сначала происходит сортировка по `Churn` в порядке неубывания, а в рамках одинаковых элементов по `Churn` происходит сортировка по `Total day charge` в порядке неубывания.

In [18]:
df.sort_values(by=["Churn", "Total day charge"], ascending=[True, False]).head()

State  Account length  Area code International plan Voice mail plan  \
688     MN              13        510                 No             Yes   
2259    NC             210        415                 No             Yes   
534     LA              67        510                 No              No   
575     SD             114        415                 No             Yes   
2858    AL             141        510                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
688                      21              315.6              105   
2259                     31              313.8               87   
534                       0              310.4               97   
575                      36              309.9               90   
2858                     28              308.0              123   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
688              53.65              208.9               71             17.76   
2259             53.35              147.7              103             12.55   
534              52.77               66.5              123              5.65   
575              52.68              200.3               89             17.03   
2858             52.36              247.8              128             21.06   

      Total night minutes  Total night calls  Total night charge  \
688                 260.1                123               11.70   
2259                192.7                 97                8.67   
534                 246.5                 99               11.09   
575                 183.5                105                8.26   
2858                152.9                103                6.88   

      Total intl minutes  Total intl calls  Total intl charge  \
688                 12.1                 3               3.27   
2259                10.1                 7               2.73   
534                  9.2                10               2.48   
575                 14.2                 2               3.83   
2858                 7.4                 3               2.00   

      Customer service calls  Churn  
688                        3      0  
2259                       3      0  
534                        4      0  
575                        1      0  
2858                       1      0

### Рассмотрим примеры

#### Каковы средние значения числовых характеристик для постоянных пользователей?

Операция маскирования - очень удобный инструмент когда нам нужно отфильтровать данные по определенному критерию. В данном случае `df["Churn"] == 1` - маска по которой будет фильтроваться набор `df`.

In [45]:
df[df["Churn"] == 1].head(5)

State  Account length  Area code  International plan  Voice mail plan  \
10    IN              65        415               False            False   
15    NY             161        415               False            False   
21    CO              77        408               False            False   
33    AZ              12        408               False            False   
41    MD             135        408                True             True   

    Number vmail messages  Total day minutes  Total day calls  \
10                      0              129.1              137   
15                      0              332.9               67   
21                      0               62.4               89   
33                      0              249.6              118   
41                     41              173.1               85   

    Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
10             21.95              228.5               83             19.42   
15             56.59              317.8               97             27.01   
21             10.61              169.9              121             14.44   
33             42.43              252.4              119             21.45   
41             29.43              203.9              107             17.33   

    Total night minutes  Total night calls  Total night charge  \
10                208.8                111                9.40   
15                160.6                128                7.23   
21                209.6                 64                9.43   
33                280.2                 90               12.61   
41                122.2                 78                5.50   

    Total intl minutes  Total intl calls  Total intl charge  \
10                12.7                 6               3.43   
15                 5.4                 9               1.46   
21                 5.7                 6               1.54   
33                11.8                 3               3.19   
41                14.6                15               3.94   

    Customer service calls  Churn  
10                       4      1  
15                       4      1  
21                       5      1  
33                       1      1  
41                       0      1

`select_dtypes` - позволяет выбрать из набора данных столбцы с определенным типом данных

In [46]:
df.select_dtypes(include=np.number)[df["Churn"] == 1].mean()

Account length            102.66
Area code                 437.82
Number vmail messages       5.12
Total day minutes         206.91
Total day calls           101.34
Total day charge           35.18
Total eve minutes         212.41
Total eve calls           100.56
Total eve charge           18.05
Total night minutes       205.23
Total night calls         100.40
Total night charge          9.24
Total intl minutes         10.70
Total intl calls            4.16
Total intl charge           2.89
Customer service calls      2.23
Churn                       1.00
dtype: float64

#### Сколько времени (в среднем) постоянные пользователи проводят на телефоне в дневное время?

Опять же применим операцию маскирования чтобы выделить постоянных пользователей, а затем выберем нужный нам столбец и посчитаем среднее

In [54]:
df[df["Churn"] == 1]["Total day minutes"].mean()

206.91407867494823

#### Какова максимальная продолжительность международных звонков среди постоянных пользователей (`Churn == 0`), у которых нет международного тарифного плана?

Здесь нам нужна фильтрация по двум критериям: `Churn == 0` - постоянный пользователь и `International plan == No` - отсутствие международного плана.

In [53]:
df[(df["Churn"] == 0) & (df["International plan"] == "No")]["Total intl minutes"].max()

18.9

### Индексация

Набор данных может быть проиндексирован как по строкам, так и по столбцам. Для этого используется метод `.loc(row range, col range)`

In [22]:
df.loc[0:5, "State":"Area code"]

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415
5    AL             118        510

In [23]:
df.iloc[0:5, 0:3]

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415

Также совершенно спокойно можно выполнять простые срезы по строкам не обращаясь к методу `.loc`

In [24]:
df[-1:]

State  Account length  Area code International plan Voice mail plan  \
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3332             39.85              265.9               82              22.6   

      Total night minutes  Total night calls  Total night charge  \
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
3332                13.7                 4                3.7   

      Customer service calls  Churn  
3332                       0      0

Также очень просто можно применить к набору данных преобразование в виде функции одного аргумента(так называемый Mapping). Функция должна иметь 1 входной аргумент(текущее значение) и 1 выходной аргумент(результат обработки)

In [25]:
df.apply(np.max)

State                        WY
Account length              243
Area code                   510
International plan          Yes
Voice mail plan             Yes
Number vmail messages        51
Total day minutes         350.8
Total day calls             165
Total day charge          59.64
Total eve minutes         363.7
Total eve calls             170
Total eve charge          30.91
Total night minutes       395.0
Total night calls           175
Total night charge        17.77
Total intl minutes         20.0
Total intl calls             20
Total intl charge           5.4
Customer service calls        9
Churn                         1
dtype: object

In [60]:
df.head(5)

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

In [64]:
def transform(x):
    if x.dtype == np.number:
        return x * x
    return x

df.apply(transform).head(5)

/var/folders/pw/d9579qgx2sg0kpzq4pdv1gcrvjn4mx/T/ipykernel_39192/3497662165.py:2: DeprecationWarning: Converting `np.inexact` or `np.floating` to a dtype is deprecated. The current result is `float64` which is not strictly correct.
  if x.dtype == np.number:


State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25           70278.01              110   
1                     26           26114.56              123   
2                      0           59243.56              114   
3                      0           89640.36               71   
4                      0           27788.89              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0           2031.30           38966.76               99            281.57   
1            754.60           38220.25              103            276.22   
2           1712.30           14689.44              110            106.09   
3           2590.81            3831.61               88             27.67   
4            803.16           21992.89              122            159.01   

   Total night minutes  Total night calls  Total night charge  \
0             59878.09                 91              121.22   
1             64719.36                103              131.10   
2             26438.76                104               53.58   
3             38769.61                 89               78.50   
4             34931.61                121               70.73   

   Total intl minutes  Total intl calls  Total intl charge  \
0              100.00                 3               7.29   
1              187.69                 3              13.69   
2              148.84                 5              10.82   
3               43.56                 7               3.17   
4              102.01                 3               7.45   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

Также полученные результаты можно использовать для маскирования

In [26]:
df[df["State"].apply(lambda state: state[0] == "W")].head()

State  Account length  Area code International plan Voice mail plan  \
9     WV             141        415                Yes             Yes   
26    WY              57        408                 No             Yes   
44    WI              64        510                 No              No   
49    WY              97        415                 No             Yes   
54    WY              87        415                 No              No   

    Number vmail messages  Total day minutes  Total day calls  \
9                      37              258.6               84   
26                     39              213.0              115   
44                      0              154.0               67   
49                     24              133.2              135   
54                      0              151.0               83   

    Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
9              43.96              222.0              111             18.87   
26             36.21              191.1              112             16.24   
44             26.18              225.8              118             19.19   
49             22.64              217.2               58             18.46   
54             25.67              219.7              116             18.67   

    Total night minutes  Total night calls  Total night charge  \
9                 326.4                 97               14.69   
26                182.7                115                8.22   
44                265.3                 86               11.94   
49                 70.6                 79                3.18   
54                203.9                127                9.18   

    Total intl minutes  Total intl calls  Total intl charge  \
9                 11.2                 5               3.02   
26                 9.5                 3               2.57   
44                 3.5                 3               0.95   
49                11.0                 3               2.97   
54                 9.7                 3               2.62   

    Customer service calls  Churn  
9                        0      0  
26                       0      0  
44                       1      0  
49                       1      0  
54                       5      1

Если мы хотим применить некоторую таблицу в качестве преобразования значений, то можно использовать операцию `.map`

В данном случае все строки `"No"` заменятся на `False`, а `"Yes"` на `True`

In [65]:
d = {"No": False, "Yes": True}
df["International plan"] = df["International plan"].map(d)
df.head()

State  Account length  Area code  International plan Voice mail plan  \
0    KS             128        415               False             Yes   
1    OH             107        415               False             Yes   
2    NJ             137        415               False              No   
3    OH              84        408                True              No   
4    OK              75        415                True              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

Также мы можем применить более специализированную операцию `replace` для того же преобразования

In [28]:
df = df.replace({"Voice mail plan": d})
df.head()

State  Account length  Area code  International plan  Voice mail plan  \
0    KS             128        415               False             True   
1    OH             107        415               False             True   
2    NJ             137        415               False            False   
3    OH              84        408                True            False   
4    OK              75        415                True            False   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

### Группировка

Давайте посмотрим на различие по статистикам между постоянными и не постоянными пользователями

Для этого нужно применить группировку данных по столбцу `Churn` и посмотреть на определенные столбцы

In [66]:
columns_to_show = ["Total day minutes", "Total eve minutes", "Total night minutes"]

#  Группировка        Выбор столбцов   Показываем статистики(без перцентилей для упрощения)
df.groupby(["Churn"])[columns_to_show].describe(percentiles=[])

Total day minutes                                   Total eve minutes  \
                  count    mean    std  min    50%    max             count   
Churn                                                                         
0                2850.0  175.18  50.18  0.0  177.2  315.6            2850.0   
1                 483.0  206.91  69.00  0.0  217.6  350.8             483.0   

                                         Total night minutes                 \
         mean    std   min    50%    max               count    mean    std   
Churn                                                                         
0      199.04  50.29   0.0  199.6  361.8              2850.0  200.13  51.11   
1      212.41  51.73  70.9  211.3  363.7               483.0  205.23  47.13   

                            
        min     50%    max  
Churn                       
0      23.2  200.25  395.0  
1      47.4  204.80  354.9

Также мы можем указать аггрегационные функции которые мы хотим видеть

In [31]:
columns_to_show = ["Total day minutes", "Total eve minutes", "Total night minutes"]

df.groupby(["Churn"])[columns_to_show].agg(["mean", "std", "min", "max"])

Total day minutes                    Total eve minutes               \
                   mean    std  min    max              mean    std   min   
Churn                                                                       
0                175.18  50.18  0.0  315.6            199.04  50.29   0.0   
1                206.91  69.00  0.0  350.8            212.41  51.73  70.9   

             Total night minutes                      
         max                mean    std   min    max  
Churn                                                 
0      361.8              200.13  51.11  23.2  395.0  
1      363.7              205.23  47.13  47.4  354.9

Еще одна немаловажная операция - сведение таблиц. Таким образом мы можем посмотреть количество постоянных и не постоянных пользователей с международными планами или без них

In [32]:
pd.crosstab(df["Churn"], df["International plan"])

International plan  False  True 
Churn                           
0                    2664    186
1                     346    137